doc/cprover-manual/modeling-shadow-memory.md

```diff
@@ -0,0 +1,148 @@
+
+```c
+void func() {
+    // Sample local object (local variable)
+    char x = 0;
+    
+    // Shadow memory field associated with local objects.
+    __CPROVER_field_decl_local("shadow", (_Bool)1);
+}
+```
+
+### `void __CPROVER_field_decl_global(type1 field_name, SSM_value_type init_value)`
+
+Type constraints:
+
+* `type1`: string literal, such as `"field"`,
+* `SSM_value_type`: value of maximum 8 bits, signed or unsigned or `_Bool`.
+
+Ditto as above, but the field declared is associated with global objects (i.e.
+global variables or heap allocated objects).
+
+```c
+// Sample global object
+int a = 10;
+
+void func() {
+    // Shadow memory field associated with global objects.
+    __CPROVER_field_decl_global("shadow", (_Bool)0);
+}
+```
+
+### `SSM_VALUE_TYPE __CPROVER_get_field(type1 *p, type2 field_name)`
+
+Type constraints:
+
+* `SSM_VALUE_TYPE`: the type of the value that was used to intialise the field
+  with during declaration,
+* `type1 *`: a non-`void` pointer to an object of type `type1`, whose address we
+   are going to use for indexing the shadow memory component for the field
+   declared,
+* `type2`: a string literal-typed value, denoting the name of the field whose
+   value we want to retrieve, such as `"field"`.
+
+Retrieves the value (currently) associated with a SSM field of the given
+pointer. This would be either the value the field was declared to be initialised
+with, or, if there had been subsequent changes to it through a `__CPROVER_set_field`,
+the value that it was last `set` with.
+
+```c
+// Sample global object
+int a = 10;
+
+void func() {
+    // Shadow memory field (called "field") associated with global objects.
+    __CPROVER_field_decl_global("shadow", (_Bool)0);
+
+    _Bool shadow_x = __CPROVER_get_field(&a, "shadow");
+    __CPROVER_assert(shadow_x == 0, "expected success: field initialised with 0");
+    __CPROVER_assert(shadow_x == 1, "expected fail: field initialised with 0");
+}
+```
+
+### `void __CPROVER_set_field(type1 *p, type2 field_name, type3 set_value)`
+
+Type constraints:
+
+* `type1 *`: a non-`void` pointer to an object of type `type1`, whose address we
+   are going to use for indexing the shadow memory component for the field
+   declared,
+* `type2`: a string literal-typed value, denoting the name of the field whose
+   value we want to retrieve, such as `"field"`,
+* `type3`: type of the value to be set. This can be `_Bool` or of any integer
+  type signed or unsigned. Notice that if this type differs from the type the
+  field was declared with (`SSM_VALUE_TYPE` above) it will be implicitly casted
+  to it.
+
+Sets the value of an SSM field, using the name supplied by the `field name` as an
+identifier for the field and using the pointer in `p` as the address for the location
+of the value to be changed, to the given value, supplied by the third argument
+`set_value`. This value needs to be of the same type as the type of the value the
+field was originally initialised with.
```

Comment:
My understanding is that the shadow memory operates on the byte level. Therefore even if you declare the field as being 16 bits, if you pass in a value of say 300, than value would be truncated to fit in 8 bits. Is my understanding correct/incorrect and does that need to be explained here?

Reply:
No, when declaring shadow memory we should use the exact type, so < 8 bits.

The truncation happens only on `set_field` and it is explained there.